app/db/models/app/counters.py

```diff
@@ -0,0 +1,95 @@
+"""
+Schema for counters.
+
+The following section includes the necessary schema for maintaining the counts
+of different entity types. These are scoped per "data source" - however the 
+concept of "data source" is not yet implemented, see PDCT-431.
+"""
+import logging
+from enum import Enum
+import sqlalchemy as sa
+from sqlalchemy.sql import text
+from app.db.session import Base
+from sqlalchemy.orm.session import object_session
+
+
+_LOGGER = logging.getLogger(__name__)
+
+#
+# DO NOT ADD TO THIS LIST BELOW
+#
+# NOTE: These need to change when we introduce "Data source" (PDCT-431)
+ORGANISATION_CCLW = "CCLW"
+ORGANISATION_UNFCCC = "UNFCCC"
+
+
+class CountedEntity(str, Enum):
+    """Entities that are to be counted."""
+
+    Collection = "collection"
+    Family = "family"
+    Document = "document"
+    Event = "event"
+
+
+class EntityCounter(Base):
+    """
+    A list of entity counters per organisation name.
+
+    NOTE: There is no foreign key, as this is expected to change
+    when we introduce data sources (PDCT-431). So at this time a
+    FK to the new datasource table should be introduced.
+
+    This is used for generating import_ids in the following format:
+
+        <organisation.name>.<entity>.<counter>.<n>
+
+    """
+
+    __tablename__ = "entity_counter"
+    __table_args__ = (
+        sa.CheckConstraint(
+            "prefix IN ('CCLW','UNFCCC')",
+            name="prefix_allowed_orgs",
+        ),
+    )
+
+    _get_and_increment = text(
+        """
+        WITH updated AS (
+        UPDATE entity_counter SET counter = counter + 1 
+        WHERE id = :id RETURNING counter
+        )
+        SELECT counter FROM updated;
+        """
+    )
+
+    id = sa.Column(sa.Integer, primary_key=True)
+    description = sa.Column(sa.String, nullable=False, default="")
+    prefix = sa.Column(sa.String, unique=True, nullable=False)  # Organisation.name
+    counter = sa.Column(sa.Integer, default=0)
+
+    def get_next_count(self) -> str:
+        """Gets the next counter value"""
+        try:
+            db = object_session(self)
+            cmd = self._get_and_increment.bindparams(id=self.id)
+            value = db.execute(cmd).scalar()
+            db.commit()
+            return value
+        except:
+            _LOGGER.exception(f"When generating counter for {self.prefix}")
+            raise
+
+    def get_import_id(self, entity: CountedEntity, n: int = 0) -> str:
```

Comment:
relying on the rest of the code to generate the n-value seems risky, especially because clashes won't be detected until a db transaction fails...

Reply:
Good thought :thinking:

Reply:
ok - thinking about this - I believe this code is fine - as this generates a new import_id - I believe what we need is another function that generates the next valid n-value.

Reply:
I think I see what you're saying here & agree with you when it comes to adding families. Perhaps we need to either move away from the structure of import IDs we have at the moment for things like documents (which currently share a 3rd quad with the family), and just always assign 0 to the 4th quad (and only to match the current import_id checking). That way anything generated by the new code is always "<source>.<type>.<guaranteed_safe_count>.0" and any implied meaning we had in previous import IDs is no longer required or expected.

I guess what I'm saying here is just "the n doesn't need to be parameterised, just always use 0" :)